operator/pkg/ciliumidentity/cache.go

```diff
@@ -310,3 +306,85 @@ func (c *CIDUsageInCES) decrementUsage(cidName int64) int {
 
 	return count
 }
+
+// EnqueueTimeTracker provides a thread safe mechanism to record
+// and manage the time when items are enqueued.
+type EnqueueTimeTracker struct {
+	enqueuedAt map[string]time.Time
+	mu         lock.Mutex
+}
+
+func (e *EnqueueTimeTracker) SetEnqueueTimeIfNotSet(item string) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if e.enqueuedAt[item].IsZero() {
+		e.enqueuedAt[item] = time.Now()
+	}
+}
+
+func (e *EnqueueTimeTracker) GetEnqueueTimeAndReset(item string) (time.Time, bool) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	enqueuedTime, exists := e.enqueuedAt[item]
+	if !exists {
+		return time.Time{}, false
+	}
+
+	delete(e.enqueuedAt, item)
+	return enqueuedTime, true
+}
+
+// CIDDeletionTracker tracks which CIDs are marked for deletion.
+// This is required for simultaneous CID management
+// by both cilium-operator and cilium-agent.
```

Comment:
Is it possible to expand this comment explaining why is this required? It would be nice to detail the expected sequence of events that proves the tracker is needed when both operator and agent manage CIDs.

Reply:
Updated the docs

Reply:
I'm sorry but I'm not following the comment.

AFAIU we need the delay (and thus the marking and deletion-tracking) to avoid scenarios where the agent created the CID but it does not have updated the CEP with the CID yet. If the operator sees the new CID and no CEP actually using it, it might be too aggressive and delete the CID. This is supposed to happen more frequently in case of high pod churn (I guess because this increases the delay between the operator receiving the CID creation event and the operator receiving the CEP update event).
Is my understanding correct? Either way, I suggest to rephrase the comment to be more clear and descriptive.

Reply:
It address two main concerns:

1. It facilitates CID reuse in scenarios with high pod churn. We are not deleting the CID and we are able to re-use it.
2. Ensure correct CID association in scenarios where the agent rapidly receives create, delete events. 
e.g. we have 1 CID used by one pod, and we churn this pod (agents quickly receives create, delete CID events).  If the agent receives the delete CID, schedules the CEP deletion, then a new create event comes, sees that the CEP is already existing, the scheduled CEP event deletes the CEP.

Reply:
@pippolo84, added more details in the comment. I think once we have the full implementation it will provide enough context.

Reply:
Thanks! Not completely sure about the wording yet, but it can be just me. @marseel WDYT? I know you have context from the CfP, any advice? :pray:

Reply:
Hey @pippolo84,
Thanks for the feedback, so one concrete example could be during the upgrade, operator might delete an CID that is set locally on CEP without the delayed delete.
Which we could reproduced with namespace label changes, when operator creates a new identity for the new labels, and agent creates a new CID for the same labels, then before operator receives the CEP with the populated CID it thinks this new CID is not used, and it deletes the new CID that was created by agent.

I'm happy to address any other requests to wording even after the further PRs will be merged, do you think we can unblock this and keep the wording discussion offline?

Reply:
I think I kind of understand the intent of it, but I'm not sure if it's required. It seems to be related to https://github.com/cilium/cilium/pull/19936

The current comment in the PR is definitely misleading as it refers to network policy as CEP? 

Regardless, AFAIK Identities are currently only deleted by the operator and the current timeout is 30 minutes + 15 minutes. 
Identitiy is considered alive if any CEP points to CID.
Identitity is considered not live **and** ready for GC if it was not alive for 30 minutes. In this case, identity is marked for deletion using annotation and the next iteration of GC is deleting it if any of the nodes did not remove annotation. 

So essentially deletion should only happen if:
- Identity was not used by any CEP for 30 minutes (2 * defaults.KVstoreLeaseTTL) - it is marked for deletion at this point
- Identity mark was not removed by any of the nodes for following 15 minutes. (defaults.KVstoreLeaseTTL)

Also, briefly looking at full implementation in https://github.com/cilium/cilium/pull/33204
I don't see any use of it too :thinking:?

Reply:
> Which we could reproduced with namespace label changes, when operator creates a new identity for the new labels, and agent creates a new CID for the same labels, then before operator receives the CEP with the populated CID it thinks this new CID is not used, and it deletes the new CID that was created by agent.

I think my comment still applies to your example case. Unless you plan to rework how GC works, I don't think "deletes the new CID that was created by agent." this can happen due to timeouts I mentioned in my comments. It would happen only after ~45 minutes by default.

Reply:
okay, everything makes sense to me. Thanks for explanation :100: 

For the wording part I would take https://github.com/cilium/cilium/pull/19936/files#diff-60a9c671858c5d13fda0333139099edf896a6e3371dc38de7df250a259c6e9a4R113-R125 as an example how to document very similar use-case. "high pod churn" seems unrelated to what it's trying to achieve.

Reply:
Although, looking at CFP it seems to me like there were changes to GC proposed, but not implementing fully new GC :thinking: ?
https://docs.google.com/document/d/1Hcc_2mB9OOUxrqQgZ-gSYDPnLYE_If_TCzVbUGDOdGM/edit
It sounds to me that new implementation will be very similar to old one.

Reply:
reworking the existing GC would too much complexity to the existing implementation and branching. Decoupling it and having the GC in the operator managing CIDs adds a clear separation. Also we need to modify the internal state of the CID (operator managing CIDs) which would've added additional complexity in re-working the existing implementation.  But indeed, the new GC should be fairly similar to the existing one.

Reply:
I also consider the delayed deletion is useful in high pod churn scenarios as we do not have to delete and re-create the same CID over and over..

Reply:
Thanks for all the explanation. I will defer to @joestringer if implementing a new GC in this case makes sense or not. 

My gut feeling is that implementing a separate GC will make it actually more complex - especially during migration from "current state" to "operator managing CIDs". GC in operator seems to me totally unrelated to who is managing CIDs.
CFP part of "CID GC reworked" sounds like the implementation of this could happen even if operator does not manage CIDs by introducing this `DeletionTracker` and replacing Heartbeat annotation with it.

Reply:
> I also consider the delayed deletion is useful in high pod churn scenarios as we do not have to delete and re-create the same CID over and over..

sure, but that's already implemented in the current GC.

Reply:
Update the docs.

@dlapcevic do you have any thoughts around the existing GC and the one proposed to be in operator managing CIDs?

Reply:
I will try to sum it up and let me know if my understanding is correct so @joestringer and @dlapcevic  doesn't have to read all of that conversation :sweat_smile: 

We are considering how GC behaves in two different cases:
(1) During the transition period from cilium-agent creating CIDs to operator creating CIDs when both operator and agent can create CIDs
(2) How GC behaves once operators is fully managing CIDs

Let's think about the current GC and how it would behave:
(1) if CID was ever used, it will be marked for deletion after 30 minutes of not being used with annotation and deleted after another 15 minutes - this already addresses your concern with high-churn pods. 
If CID was never used by any CEP, it will mark it for deletion and delete it after 15 minutes.
Sounds reasonable both for identities created by the operator and the agent.
CIDs created by the agent and not yet used by any CEPs won't be deleted immediately. 
 
(2) It will still work the same, but adding an annotation mark for deletion won't make much sense as the agent won't be able to remove it anymore.
So proposed switch from `deletion mark annotation` to `DeletionTracker` is a nice optimization in terms of API calls, but not necessary for it to work correctly. It should be fairly easy to implement it in the current GC though.  


It could probably even be instantly enabled ( `deletion mark annotation` -> `DeletionTracker`) when switching from agent creating CIDs to operator creating CIDs during migration and the current GC would still behave correctly in both cases (1) and (2), but would not react to deletion mark annotation, but used fixed delay for waiting on CEP.

Of course interval/timeouts (30/15 minutes) can be already changed IIRC, I was just using defaults as examples. 


So the question is, what would be the purpose of implementing a new GC, instead of implementing this improvement to the current GC, which seems significantly easier?

Reply:
In my opinion, GC is part of CID management, and there is no reason for it to be separated from the CID controller. The CID controller anyway needs to know the desired state of CIDs, when CIDs are used and when not, and adding deletion to it is not a complex effort.

Beside that I see a few benefits:
1. Performance and scalability\
a) Smaller number of calls to kube-apiserver without adding and removing the mark. Remember: every update needs to be sent to all nodes.
b) Processing all CIDs at one time (each GC run) instead of continuously, creates a spike of requests that can significantly affect KCP and cilium-operator's k8s client’s rate limiting.
2. Ease of use\
No need to configure GC.
3. Security\
Remove write permission for CIDs from cilium-agent.\
When cilium-agent is not managing CIDs, it shouldn’t update CIDs, so it won’t need write permission to CIDs.

We had issues related to these points.
We also had issues where cilium-operator was having connection instability with kube-apiserver so it was restarting every 15-20 minutes (or leader was changed), and in this case we ended up having no CIDs cleaned up and eventually hitting the 65k CID limit.

Reply:
I think that this topic would be a lot more approachable with a more incremental approach. Use the current GC first, propose improvements to the current GC implementation and push on that independently. Then propose the PR that refactors/removes the current GC implementation to merge it into this CID controller. Each of those steps would be self-contained, provide value, and could be individually assessed for their benefits and drawbacks.

If we were looking at a PR that was titled "Improve cilium-operator Identity garbage collection" with the PR description of the comment immediately above this comment, then I think we would be in a much better position to critique the design and implementation, and debate how it affects the implementation.

Given how critical these operations are and how nuanced they can be, I don't think we're interested in having multiple implementations in the tree at the same time. Not unless there's something critical to the design that I'm missing which makes the garbage collection improvements inherently different.